.github/workflows/main.yaml

```diff
@@ -16,68 +16,42 @@ jobs:
+        uses: docker/setup-buildx-action@v2
       - name: Extract targets
         run: |
-          TAG_MINIMAL=$(cat TAG_MINIMAL)
-          c="$(../tag_exists ubuntu-minimal $TAG_MINIMAL)"
-          if [ "$c" = ng ]; then
-              echo "update ubuntu-minimal:$TAG_MINIMAL"
-              echo "yes!" > NEED_UPDATE_MINIMAL
-          fi
           TAG=$(cat TAG)
-          for image in ubuntu ubuntu-dev ubuntu-debug; do
+          for image in ubuntu-minimal ubuntu ubuntu-dev ubuntu-debug; do
               c="$(../tag_exists $image $TAG)"
               if [ "$c" = ng ]; then
                   echo "build $image:$TAG"
                   echo $image >> BUILDS
               fi
           done
-      - name: Pull minimal image
-        run: |
-          if [ ! -f NEED_UPDATE_MINIMAL ]; then
-              echo "nothing to pull."
-              exit 0
-          fi
-          TAG_MINIMAL=$(cat TAG_MINIMAL)
-          echo
-          echo "pulling ubuntu-minimal ..."
-          docker pull ubuntu:$TAG_MINIMAL
-          docker tag ubuntu:$TAG_MINIMAL quay.io/cybozu/ubuntu-minimal:$TAG_MINIMAL
-      - name: Push minimal image
+      - name: Build and push images
         if: github.event_name != 'pull_request'
-        run: |
-          if [ ! -f NEED_UPDATE_MINIMAL ]; then
-              exit 0
-          fi
-          echo ${{ secrets.QUAY_PASSWORD }} | docker login -u ${{ secrets.QUAY_USER }} --password-stdin quay.io
-          TAG_MINIMAL=$(cat TAG_MINIMAL)
-          echo "pushing ubuntu-minimal ..."
-          docker push quay.io/cybozu/ubuntu-minimal:$TAG_MINIMAL
-      - name: Build images
         run: |
           if [ ! -f BUILDS ]; then
               echo "nothing to build."
               exit 0
           fi
-          TAG_MINIMAL=$(cat TAG_MINIMAL)
-          for d in $(cat BUILDS); do
-              echo
-              echo "building $d ..."
-              docker build -t quay.io/cybozu/${d}:latest --build-arg TAG_MINIMAL=$TAG_MINIMAL ${d}
-          done
-      - name: Push docker images
-        if: github.event_name != 'pull_request'
-        run: |
-          if [ ! -f BUILDS ]; then
-              exit 0
-          fi
           echo ${{ secrets.QUAY_PASSWORD }} | docker login -u ${{ secrets.QUAY_USER }} --password-stdin quay.io
+          TAG_MINIMAL=$(cat TAG_MINIMAL)
           TAG=$(cat TAG)
           BRANCH=$(cat BRANCH)
           for d in $(cat BUILDS); do
-              echo "pushing ${d} ..."
-              docker tag quay.io/cybozu/${d}:latest quay.io/cybozu/${d}:$TAG
-              docker push quay.io/cybozu/${d}:$TAG
-              docker tag quay.io/cybozu/${d}:latest quay.io/cybozu/${d}:$BRANCH
-              docker push quay.io/cybozu/${d}:$BRANCH
+              echo
+              echo "building $d ..."
+              docker buildx build \
+                --platform linux/amd64,linux/arm64/v8 \
+                --push \
+                -t quay.io/cybozu/${d}:latest \
```

Comment:
We don't want the `latest` tag to be pushed.
This job is executed for both `18.04` and `20.04`, so `latest` becomes ambiguous.

`{18.04,20.04}/ubuntu-{debug,dev}/Dockerfile` use `FROM quay.io/cybozu/ubuntu:latest`.
The original code expects the `latest` tag to be created (but not pushed) in the loop.
Because we tag and push in one command, it would be good to remove use of `latest` by modifying `FROM quay.io/cybozu/ubuntu:latest` to `FROM quay.io/cybozu/ubuntu:${TAG}`.

Reply:
Sorry, I fixed to don't push the `latest` tag.
And I removed use of `latest` by modifying `FROM quay.io/cybozu/ubuntu:latest` to `FROM quay.io/cybozu/ubuntu:${TAG}` as you pointed out.